### ISSUE TITLE
Avoid `expectFail` in the test suite

### ISSUE THREAD

**Issue by fendor** [haskell/haskell-language-server#4130]
[`expectFail`](https://hackage.haskell.org/package/tasty-expected-failure-0.12.3/docs/Test-Tasty-ExpectedFailure.html#v:expectFail) interprets a test failure for any reason as a success. But since we don't check why exactly the test is failing, we lose valuable information.
Thus, many of the tests that use `expectFail` are likely not really testing what they are supposed to be.

In my opinion, we should replace all occurrences of `expectFail` with an assertion that shows what is failing precisely, or delete/ignore the respective test case.

**Comment by sgillespie** [comment]
Most of the `expectFail` usages are to "disable" tests known to be broken. Is this a valid use, or should they all be updated/removed?

**Comment by fendor** [comment]
I think `expectFail` is fine to temporarily disable certain tests on certain platforms.
The issue is with unconditionally disabling a test with `expectFail`, as we lose over time the information *why* the test is supposed to be failing. In these situation, as long as it isn't too flaky, we should have an assertion that documents the current behviour, with a comment explaining what is going wrong!

**Comment by sgillespie** [comment]
I updated a few tests in #4402 for discussion. Is this basically what you were imagining?

**Comment by fendor** [comment]
Yes, that's basically what I am imagining, thanks!
I think it would be nice to also document what the intended behaviour would be, preferably in the code. E.g., adding a `hoverTestExpectedFail` which takes the expected but incorrect value, and also the *really* expected value as input parameters. 

Do you think that would make sense?

**Comment by sgillespie** [comment]
I like the idea of correctly typing both possibilities. Maybe oversimplifying, but something like this?
```
hoverTestExpectedFail :: Assertion -> Assertion -> Assertion
hoverTestExpectedFail expectedIncorrect _expectedIntended = expectedIncorrect
```

**Comment by fendor** [comment]
This looks nice to me!

**Comment by sgillespie** [comment]
> This looks nice to me!

Which one?

**Comment by fendor** [comment]
Both look good to me, the second proposal sounds better and I'd be curious what the results look like in practice.

**Comment by sgillespie** [comment]
I've updated most of the tests. I'm not really sure what to do with the golden tests, so I left those alone. Otherwise, they all make use of a simple GADT that's meant to distinguish between current and ideal behaviors.

**Comment by fendor** [comment]
Thanks, closed by #4402
